loki/transform/transform_utilities.py

```diff
@@ -144,6 +144,36 @@ def replace_intrinsics(routine, function_map=None, symbol_map=None, case_sensiti
     routine.spec = SubstituteExpressions(callmap).visit(routine.spec)
     routine.body = SubstituteExpressions(callmap).visit(routine.body)
 
+def rename_variables(routine, symbol_map=None):
+    """
+    Replace symbols/variables including (routine) arguments.
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine object in which to replace intrinsic calls
+    symbol_map : dict[str, str]
+        Mapping from symbol/variable names to their replacement
+    """
+    symbol_map = CaseInsensitiveDict(symbol_map) or {}
+    # rename arguments if necessary
+    arguments = ()
+    for arg in routine.arguments:
+        if arg.name in symbol_map:
+            arguments += (arg.clone(name=symbol_map[arg.name]),)
+        else:
+            arguments += (arg,)
+    routine.arguments = arguments
+    # rename variable declarations and usages
+    var_map = {}
+    for var in FindVariables(unique=False).visit(routine.ir):
+        if var.name in symbol_map:
+            new_var = symbol_map[var.name]
+            if new_var is not None:
+                var_map[var] = var.clone(name=symbol_map[var.name])
+    if var_map:
+        routine.spec = SubstituteExpressions(var_map).visit(routine.spec)
+        routine.body = SubstituteExpressions(var_map).visit(routine.body)
```

Comment:
Maybe we should update here also the symbol table to make sure we don't have leftover entries under the previous name? Something along the lines of
```py
for name in var_map:
    if name in routine.symbol_attrs:
        del routine.symbol_attrs[name]
```

Reply:
Yes! Suggested solution not working for `DerivedType` s.